roxie/ccd/ccdserver.cpp

```diff
@@ -11835,10 +10994,25 @@ class CRoxieServerJoinActivity : public CRoxieServerTwoInputActivity
             createDefaultLeft();
         if ((leftOuterJoin && (activityKind==TAKjoin || activityKind==TAKjoinlight || activityKind==TAKdenormalizegroup)) || limitOnFail)
             createDefaultRight();
-        if ((helper.getJoinFlags() & JFlimitedprefixjoin) && helper.getJoinLimit()) 
+        bool isStable = (helper.getJoinFlags() & JFunstable) == 0;
+        RoxieSortAlgorithm sortAlgorithm;
+        if (forceSpill)
+            sortAlgorithm = isStable ? stableSpillingQuickSortAlgorithm : spillingQuickSortAlgorithm;
+        else
+            sortAlgorithm = isStable ? stableQuickSortAlgorithm : quickSortAlgorithm;
+        if (helper.isLeftAlreadySorted())
+            groupedLeft.setown(createGroupedInputReader(input, NULL)); // This is something of a no-op but helps keep the code simple (IInputBase is not linkable)
+        else
+            groupedLeft.setown(createSortedGroupedInputReader(input, NULL, createSortAlgorithm(sortAlgorithm, helper.queryCompareLeft(), ctx->queryRowManager(), input->queryOutputMeta(), ctx->queryCodeContext(), tempDirectory, activityId)));
+        ICompare *compareRight = helper.queryCompareRight();
+        if (helper.isRightAlreadySorted())
+            groupedRight.setown(createGroupedInputReader(input1, compareRight));
+        else
+            groupedRight.setown(createSortedGroupedInputReader(input1, compareRight, createSortAlgorithm(sortAlgorithm, compareRight, ctx->queryRowManager(), input1->queryOutputMeta(), ctx->queryCodeContext(), tempDirectory, activityId)));
```

Comment:
Creating the sort algorithm could be done once in the onCreate.  May be significant if you have joins inside a child query.